nntrainer/graph/graph_node.h

```diff
@@ -74,5 +75,228 @@ class GraphNode {
   // virtual GraphNode &copy(const GraphNode &from) = 0;
 };
 
+/**
+ * @brief   Iterator for GraphNode which return LayerNode object upon realize
+ *
+ * @note    This does not include the complete list of required functions. Add
+ * them as per need.
+ */
+template <typename T>
+class GraphNodeIterator : public std::iterator<std::random_access_iterator_tag,
+                                               std::shared_ptr<GraphNode>> {
+  std::shared_ptr<GraphNode> *p; /** underlying object of GraphNode */
+
+public:
+  /**
+   * @brief   iterator_traits types definition
+   *
+   * @note    these are not requried to be explicitly defined now, but maintains
+   *          forward compatibility for c++17 and later
+   *
+   * @note    valute_type, pointer and reference are different from standard
+   * iterator
+   */
+  typedef std::shared_ptr<T> value_type;
+  typedef std::random_access_iterator_tag iterator_category;
+  typedef std::ptrdiff_t difference_type;
+  typedef std::shared_ptr<T> *pointer;
+  typedef std::shared_ptr<T> &reference;
+
+  /**
+   * @brief Construct a new Graph Node Iterator object
+   *
+   * @param x underlying object of GraphNode
+   */
+  GraphNodeIterator(std::shared_ptr<GraphNode> *x) : p(x) {}
+
+  /**
+   * @brief reference operator
+   *
+   * @return value_type
+   * @note this is different from standard iterator
+   */
+  value_type operator*() const { return std::static_pointer_cast<T>(*p); }
```

Comment:
return `reference_type`?

Reply:
> and `*std::static_pointer_cast<T>(*p)`

This will return `T` type and not `std::shared_ptr<T>`. Please check this once with the last commit of the PR which has the the final version of the GraphNodeIterator.

Reply:
Acked